### ISSUE TITLE
AL2 - No usable version of libssl was found

### ISSUE THREAD

**Issue by KMacTN** [devizer/w3top-bin#3]
First of all, w3top is awesome! However, in Amazon Linux 2 it installs fine but isn't about to start up.  It is not finding the openssl libraries it wants and core dumps with a message "No usable version of libssl was found".  It looks like it is wanting openssl 1.0, which Amazon has removed from their repos.  Is there a workaround or fix for this?

**Comment by devizer** [comment]
Two openssl versions are supported - 1.0.2 and 1.1. v3 is not supported, but 1.0.2, 1.1, and 3.0 can be installed side by side because the names of so-files are different. I tested it on Amazon Linux v2 in 2020. AFAIK, It was sufficient to 
```bash
sudo yum install openssl11
# or
sudo yum install curl
```

Please let me know if it works for you.

**Comment by KMacTN** [comment]
Openssl11 worked, ty!-KevinOn Jun 14, 2023, at 6:10 PM, devizer ***@***.***> wrote:﻿
Two openssl versions are supported - 1.02 and 1.1. v3 is not supported, but 1.0.2, 1.1, and 3.0 can be installed side by side because the names of so-files are different. I tested it on Amazon Linux v2 in 2020. AFAIK, It was sufficient to
sudo yum install openssl11
# or
sudo yum install curl
Please let me know if it works for you.

**Comment by speedtoys** [comment]
I downloaded and compiled openssl-1.1.1 for Ubuntu 22.04 here, and the make install was a success.

But I still fail with:

Sep 27 18:48:32 speedplex systemd[1]: Stopped W3Top service.
Sep 27 18:48:32 speedplex systemd[1]: w3top.service: Consumed 1.410s CPU time.
Sep 27 18:48:32 speedplex systemd[1]: Started W3Top service.
Sep 27 18:48:32 speedplex w3top[1949726]: Runtime: .NET Core 3.1.26
Sep 27 18:48:32 speedplex w3top[1949726]: PID file for Universe.W3Top.dll ver 1.27.2576.210 created: /var/run/w3top.pid
Sep 27 18:48:32 speedplex w3top[1949726]: System Environment:
Sep 27 18:48:32 speedplex w3top[1949726]:     OS ................ Ubuntu 22.04.3 LTS, Kernel 5.15.0-83-generic (x86_64)
Sep 27 18:48:32 speedplex w3top[1949726]:     CPU ............... Intel(R) Xeon(R) CPU E5-2690 v2 @ 3.00GHz, Cache 25600 KB, 40 Cores
Sep 27 18:48:32 speedplex w3top[1949726]:     Memory ............ 257,882 MB; Working Set: 24 MB
Sep 27 18:48:32 speedplex w3top[1949726]: Stopwatch #1: ProcMountsParser(/proc/mounts) successfully pre-jitted in 56.422 msec (cpu: 88%, 49.402 = 49.402 [user] + 0.000 [kernel] milliseconds)
Sep 27 18:48:32 speedplex w3top[1949726]: Stopwatch #2: LinuxMemorySummary.TryParse(/proc/meminfo) successfully pre-jitted in 4.636 msec (cpu: 86%, 3.984 = 3.984 [user] + 0.000 [kernel] milliseconds)
Sep 27 18:48:32 speedplex w3top[1949726]: Stopwatch #3: SysBlocksReader(/sys/block) successfully pre-jitted in 19.735 msec (cpu: 101%, 19.944 = 19.944 [user] + 0.000 [kernel] milliseconds)
Sep 27 18:48:32 speedplex w3top[1949726]: Universe.Dashboard.DAL SqLite DB: /root/.cache/W3Top/history-1.0.0.15.sqlite
Sep 27 18:48:32 speedplex w3top[1949726]: Storage FAMILY: Sqlite
Sep 27 18:48:33 speedplex w3top[1949726]: Stopwatch #4: Create/Upgrade Sqlite DB Structure in 130.817 msec (cpu: 101%, 131.999 = 128.003 [user] + 3.996 [kernel] milliseconds)
Sep 27 18:48:33 speedplex w3top[1949726]: #033[40m#033[1m#033[33mwarn#033[39m#033[22m#033[49m: Microsoft.EntityFrameworkCore.Model.Validation[10400]
Sep 27 18:48:33 speedplex w3top[1949726]:       Sensitive data logging is enabled. Log entries and exception messages may include sensitive application data, this mode should only be enabled during development.
Sep 27 18:48:33 speedplex w3top[1949726]: DB Server is ready. Its version is: 3.33.0
Sep 27 18:48:33 speedplex w3top[1949726]: Skipping Broadcasting: Services are not yet injected
Sep 27 18:48:33 speedplex w3top[1949726]: Stopwatch #5: Prepare shared /proc/mounts data source in 8.820 msec (cpu: 5%, 0.423 = 0.404 [user] + 0.019 [kernel] milliseconds)
Sep 27 18:48:33 speedplex w3top[1949726]: Stopwatch #6: Prepare shared /proc/swaps data source in 7.267 msec (cpu: 6%, 0.436 = 0.436 [user] + 0.000 [kernel] milliseconds)
Sep 27 18:48:33 speedplex w3top[1949726]: No usable version of libssl was found
Sep 27 18:48:34 speedplex systemd[1]: w3top.service: Main process exited, code=killed, status=6/ABRT
Sep 27 18:48:34 speedplex systemd[1]: w3top.service: Failed with result 'signal'.
Sep 27 18:48:34 speedplex systemd[1]: w3top.service: Consumed 1.404s CPU time.

**Comment by devizer** [comment]
@speedtoys, ver 1.1.1 should be ok.
Could you check up new library is properly registered by ldconfig:
`sudo ldconfig` - update cache for newly copied/compiled libraries
`ldconfig -p | grep libssl` - show available libssl libraries. On my Ubuntu 22.04 it looks like:
```
        libssl3.so (libc6,x86-64) => /lib/x86_64-linux-gnu/libssl3.so
        libssl.so.3 (libc6,x86-64) => /lib/x86_64-linux-gnu/libssl.so.3
        libssl.so.1.1 (libc6,x86-64) => /opt/networking/lib64/libssl.so.1.1
        libssl.so (libc6,x86-64) => /opt/networking/lib64/libssl.so
```

Can you see your 1.1.1 version?

**Comment by devizer** [comment]
Precompiled libssl 1.1.1m so-binaries for a set of platforms: 
```bash
# libssl 1.1.1m side-by-side binaries with ld.so.conf registration
# Special for Ubuntu 22.04, Fedora 36+, and other linux without libssl 1.1
# Supported Arch: x86_64, i386, arm, aarch64
export INSTALL_DIR=/opt/curl-temp TOOLS="curl"; script="https://master.dl.sourceforge.net/project/gcc-precompiled/build-tools/Install-Build-Tools.sh?viasf=1"; (wget -q -nv --no-check-certificate -O - $script 2>/dev/null || curl -ksSL $script) | sudo -E bash

tmp="$(mktemp)"
test -d /opt/networking/lib64 && echo /opt/networking/lib64 >> "$tmp"
test -d /opt/networking/lib && echo /opt/networking/lib >> "$tmp"
sudo cat /etc/ld.so.conf >> "$tmp"; 
sudo mv -f "$tmp" /etc/ld.so.conf
sudo ldconfig
sudo rm -rf /opt/curl-temp
```

**Comment by devizer** [comment]
@KMacTN @speedtoys, Fix 1.28: Focus on Ubuntu 22.04+, Debian 12+, Fedora 36+, etc, which are equipped with libssl 3.x+

* libssl 1.1.1m included in distribution (optional/libssl-1.1 folder) for armv7, arm64 and x86_64. 
* System libssl 1.1.* is preferred if installed when you use install-w3top-service.sh or install-systemd-service.sh
* If you use a custom Universe.W3Top launcher, then libssl path can be passed by LD_LIBRARY_PATH 
`LD_LIBRARY_PATH=$(pwd)/optional/libssl-1.1 ./Universe.W3Top` or /etc/ld.so.conf

Thanks for the feedback
